### ISSUE TITLE
Does adding trailing zeros affect round-trip?

### ISSUE THREAD

**Issue by expnkx** [jk-jeon/Grisu-Exact#8]
For example. double has 17 digits.
23535782139572189357.0

Which one is the correct answer? I just add trailing zeros after the scientific part for general (pick the shorter one between scientific and fixed).
fast_io:23535782139572190000
microsoft STL charconv:23535782139572191232

**Comment by jk-jeon** [comment]
It should not affect roundtrip.
But the problem is the resulting output is probably not what the user want

**Comment by jk-jeon** [comment]
It should not affect roundtrip.
But the problem is that the result is not what the user probably want.
Note that the whole point of providing a way to indicate precision is not to be able beautifully align the output string; it is to give the user an ability to choose the precision they are working with.
Ignoring that and just putting trailing zeros is a cheating at best.

And also, please stop abusing issue pages as your personal Q&A section. I can't see anything related to Grisu-Exact in your question.

**Comment by expnkx** [comment]
> It should not affect roundtrip.
> But the problem is that the result is not what the user probably want.
> Note that the whole point of providing a way to indicate precision is not to be able beautifully align the output string; it is to give the user an ability to choose the precision they are working with.
> Ignoring that and just putting trailing zeros is a cheating at best.
> 
> And also, please stop abusing issue pages as your personal Q&A section. I can't see anything related to Grisu-Exact in your question.

Ty. Can you provide me your email so I could ask you for these questions? Since I need more basic understanding of these questions so I might help you.

**Comment by expnkx** [comment]
https://bitbucket.org/ejsvifq_mabmip/fast_io/src/reserver_test/testsuites/0007.round_trip/fp.cc
Look at this. No precision is required here.

**Comment by jk-jeon** [comment]
> I know there is a difference between the shortest round-trip general/fixed/scientific and the precision general/fixed/scientific. However, the shortest round-trip general/fixed/scientific is not something CRT provides. For example, %g for printf or std::cout<<1.0 only provides you 6 digits general precisions by default.
>Is it something new since ryu and grisu-exact? I do not know.

The idea of shortest output with roundtrip guarantee dates back to Steele Jr. and White's paper _How to print floating-point numbers accurately_ in 1990. As you know, `printf` and `iostream` are much older than that, although Steele Jr. and White were probably not the first who came up with the idea. As far as I know, their proposed algorithm, called Dragon4, was not extremely popular; it was Florian Loitsch's new algorithm which he called Grisu2 & Grisu3, presented in 2010, which really made the idea of shortest output super popular. Probably JavaScript and JSON's wide adaptation was also an important factor. So, yes, it is sort of new.

I don't know how you wish to handle precision settings in your library, but as you have noticed, shortest output mode and the fixed precision mode are completely incompatible and orthogonal modes. Ryu also offers two different algorithms for those: Ryu and Ryu-printf; those are two (related but) completely different algorithms.

To me, indicating the "precision" when using the shortest output mode is a nonsense. Also, using trailing zeros with the shortest output mode is even more nonsense. Well, you might want nicely aligned output even with the shortest mode, but no, don't do that, because it suggests that those digits are precise. This is the standard convention used by scientists/engineers/mathematicians/etc..

tl;dr
You should separate fixed-precision mode from shortest output mode, and shouldn't ever try to mix them.

**Comment by expnkx** [comment]
> > I know there is a difference between the shortest round-trip general/fixed/scientific and the precision general/fixed/scientific. However, the shortest round-trip general/fixed/scientific is not something CRT provides. For example, %g for printf or std::cout<<1.0 only provides you 6 digits general precisions by default.
> > Is it something new since ryu and grisu-exact? I do not know.
> 
> The idea of shortest output with roundtrip guarantee dates back to Steele Jr. and White's paper _How to print floating-point numbers accurately_ in 1990. As you know, `printf` and `iostream` are much older than that, although Steele Jr. and White were probably not the first who came up with the idea. As far as I know, their proposed algorithm, called Dragon4, was not extremely popular; it was Florian Loitsch's new algorithm which he called Grisu2 & Grisu3, presented in 2010, which really made the idea of shortest output super popular. Probably JavaScript and JSON's wide adaptation was also an important factor. So, yes, it is sort of new.
> 
> I don't know how you wish to handle precision settings in your library, but as you have noticed, shortest output mode and the fixed precision mode are completely incompatible and orthogonal modes. Ryu also offers two different algorithms for those: Ryu and Ryu-printf; those are two (related but) completely different algorithms.
> 
> To me, indicating the "precision" when using the shortest output mode is a nonsense. Also, using trailing zeros with the shortest output mode is even more nonsense. Well, you might want nicely aligned output even with the shortest mode, but no, don't do that, because it suggests that those digits are precise. This is the standard convention used by scientists/engineers/mathematicians/etc..
> 
> tl;dr
> You should separate fixed-precision mode from shortest output mode, and shouldn't ever try to mix them.

Thank you!! Your explanation is perfect. I think this part can also be in your paper since the paper _How to Print Floating-Point Numbers Accurately_ was using zero or space for padding. So technically I am doing the exact the same thing based on that paper's approach. I believe you can make a good argument in your paper of why it is nonsense or propose something like a standard for this.
![image](https://user-images.githubusercontent.com/57303909/77143326-58a40800-6a59-11ea-8f56-bfc99735f783.png)
As you said, putting padding zeros for fixed mode is basically nonsense since it can lead to misleading for an illusion of precise value. However, it looks like the shortest round-trip one is always nonsense to somewhat degree. Just, for example, 1e-01 does not mean the binary representation is 1e-01 Since its internal value by no means a 1e-01 right? If my understanding is correct here.

The standard library charconv requires the shortest representation requirement. Does WG21 ever consider this?
TBH, I think a huge amount of nonsense is for C++17 charconv and that is why I think it is a mistake. For example, precision hexfloat? Maybe we need to submit a defect report to WG21 for this?


My strategy is basically to treat scientific as extra 4 (if e10 is smaller than 100) or 5(if e10 is larger than 100) characters as a comparison between scientific shortest format and fixed "shortest" format. The strategy code is here:
https://bitbucket.org/ejsvifq_mabmip/fast_io/src/reserver_test/include/fast_io_freestanding_impl/grisu_exact/grisu_exact_api.h

Examples:
https://bitbucket.org/ejsvifq_mabmip/fast_io/raw/058d6121f4854e23bcce8d8ae07e9b94ba087b18/examples/0003.floating_point/general.cc

**Comment by jk-jeon** [comment]
>As you said, putting padding zeros for fixed mode is basically nonsense since it can lead to misleading for an illusion of precise value. However, it looks like the shortest round-trip one is always nonsense to somewhat degree. Just, for example, 1e-01 does not mean the binary representation is 1e-01 Since its internal value by no means a 1e-01 right? If my understanding is correct here.

My opinion (although it might not be so popular) is that we should think of a bit pattern to represent an _interval_, not a single real number. Thus, although the "exact" value of a specific bit pattern might not be equal to `1e-01`, `1e-01` is certainly a "correct" output because it is in the interval corresponding to the bit pattern.

Due to intrinsic limitation of precision, we can't know what is the "true" value we might have wanted inside the interval; it might be different from the "exact" value we obtain by directly interpreting the given bit pattern as a single value. And because of [Zipf's law](https://en.wikipedia.org/wiki/Zipf%27s_law), the most probable candidate for the "true" value inside the corresponding interval is perhaps the one that is shortest. For example, although `0.29999999999999997` is closer to the "exact" value of certain bit pattern, it is highly likely that in fact the intended value was `0.3`.

In short, I think there is nothing wrong with the shortest output mode. (Also, there is nothing wrong with the fixed precision output mode neither; they are complementary things serving different use-cases.)


>My strategy is basically to treat scientific as extra 4 (if e10 is smaller than 100) or 5(if e10 is larger than 100) characters as a comparison between scientific shortest format and fixed "shortest" format.

I hope you are not confusing between "fixed-point output format" and "fixed-precision output format". By "fixed-point" I mean the position of the decimal point is fixed; for example, things like `50.1` or `0.0023` are in fixed-point format, while `1e23` or `1.32e12` are in floating-point (or scientific) format. On the other hand, by "fixed-precision" I mean the number of decimal significand digits is fixed. For example, both `0.0001230` and `5.678e10` are of precision of 4 decimal digits.

**Comment by expnkx** [comment]
> > As you said, putting padding zeros for fixed mode is basically nonsense since it can lead to misleading for an illusion of precise value. However, it looks like the shortest round-trip one is always nonsense to somewhat degree. Just, for example, 1e-01 does not mean the binary representation is 1e-01 Since its internal value by no means a 1e-01 right? If my understanding is correct here.
> 
> My opinion (although it might not be so popular) is that we should think of a bit pattern to represent an _interval_, not a single real number. Thus, although the "exact" value of a specific bit pattern might not be equal to `1e-01`, `1e-01` is certainly a "correct" output because it is in the interval corresponding to the bit pattern.
> 
> Due to intrinsic limitation of precision, we can't know what is the "true" value we might have wanted inside the interval; it might be different from the "exact" value we obtain by directly interpreting the given bit pattern as a single value. And because of [Zipf's law](https://en.wikipedia.org/wiki/Zipf%27s_law), the most probable candidate for the "true" value inside the corresponding interval is perhaps the one that is shortest. For example, although `0.29999999999999997` is closer to the "exact" value of certain bit pattern, it is highly likely that in fact the intended value was `0.3`.
> 
> In short, I think there is nothing wrong with the shortest output mode. (Also, there is nothing wrong with the fixed precision output mode neither; they are complementary things serving different use-cases.)
> 
> > My strategy is basically to treat scientific as extra 4 (if e10 is smaller than 100) or 5(if e10 is larger than 100) characters as a comparison between scientific shortest format and fixed "shortest" format.
> 
> I hope you are not confusing between "fixed-point output format" and "fixed-precision output format". By "fixed-point" I mean the position of the decimal point is fixed; for example, things like `50.1` or `0.0023` are in fixed-point format, while `1e23` or `1.32e12` are in floating-point (or scientific) format. On the other hand, by "fixed-precision" I mean the number of decimal significand digits is fixed. For example, both `0.0001230` and `5.678e10` are of precision of 4 decimal digits.

Thank you for your explanation. I think I have come up with a precise infinite length string to the floating-point algorithm. Would you like to have a talk with me about this since we could solve the reverse problem elegantly I believe? However, I am not a mathematician (like you). My idea might be flawed. However, my testing looks at proving my correctness.

**Comment by expnkx** [comment]
Here is my idea of doing this.

https://bitbucket.org/ejsvifq_mabmip/fast_io/raw/4dad94f4b645f8f95e27f88c8dfe3b2b234662f7/examples/0003.floating_point/roundtrip1.cc
https://bitbucket.org/ejsvifq_mabmip/fast_io/raw/4dad94f4b645f8f95e27f88c8dfe3b2b234662f7/examples/0003.floating_point/roundtrip2.cc

However, I do not know how the website base convert gets the precise value of them. Does it seem like Ryu precision scientific is doing this? Can we get the precise value efficiently (not the shortest round-trip one)?
https://baseconvert.com/ieee-754-floating-point

**Comment by jk-jeon** [comment]
I have no idea what is your idea.

**Comment by expnkx** [comment]
> I have no idea what is your idea.

I am finding out the peak value and then making a comparison between this and the input.

ryu sto will might offset a bit, This will fix it by verifying which one is correct.

**Comment by jk-jeon** [comment]
Well, I still can't get what's the point. Sorry:(

**Comment by expnkx** [comment]
for example any input string 
1.2342342347326444xxxx
1.2342342347326445yyyyyyy
They must have the same round-trip value no matter what values they are.

However, since the round-trip value 1.2342342347326445 is different from 1.2342342347326446.
1.2342342347326445xxxx
1.2342342347326446yyyyyyy

We take the middle point of their precise value
We take the middle value of 1.2342342347326444862432026638998650014400482177734375 and 1.2342342347326447082878075889311730861663818359375 is 1.23423423473264459726550512641551904380321502685546875

Any string which is greater than 1.23423423473264459726550512641551904380321502685546875 has the same round-trip value of 1.2342342347326446
Otherwise the round-trip value is same as 1.2342342347326445

As long as we can calculate this middle point value very fast. The problem is solved.

**Comment by jk-jeon** [comment]
You can ignore after 17th digits; rest of the digits are just used for determining the position of decimal point.

The main strategy should be this:
(1) Write your decimal number as x.x....x * 10^y...y.
(2) Take xx....x and y...y as the input, and convert it into a binary number using the same idea used in Ryu/Grisu or whatever.
(3) Pack that binary number into IEEE format.
Are you talking about something different from this?

**Comment by expnkx** [comment]
"Are you talking about something different from this?"

"You can ignore after 17th digits; rest of the digits are just used for determining the position of the decimal point."

No. This is false. I give you an example:
1.234234234732644597 :
0 01111111111 0011101111110110110001100101100111000111110000010111
1.234234234732644598 :
0 01111111111 0011101111110110110001100101100111000111110000011000

It shifts one.
That is why we need to find out the middle point between the precise value of 
0 01111111111 0011101111110110110001100101100111000111110000010111
(1.2342342347326444862432026638998650014400482177734375)
and
0 01111111111 0011101111110110110001100101100111000111110000011000
(1.2342342347326447082878075889311730861663818359375)

which is 
1.23423423473264459726550512641551904380321502685546875

**Comment by jk-jeon** [comment]
Aha. Now I can follow what you were talking about.

**Comment by expnkx** [comment]
> Aha. Now I can follow what you were talking about.

The question is about how to fast calculate the middle point and how to avoid full calculation of the middle point. Just calculate the decimal length to around the string decimal length. Then the problem is solved and I think this should be extremely fast.

I tested ryu algorithm. It runs around 0.4s per 10M double. Microsoft charconv will run 3 sec. So as long as the verification algorithm does not run 8 times slower. It will definitely be a win.

**Comment by jk-jeon** [comment]
Does Ryu-printf work only for limited length string?

>I've tested ryu sto algorithm. It runs around 0.4s per 10M double. Microsoft charconv will run at least 3 sec (sometimes it will run 7 seconds is your exp part is large enough). So as long as the verification algorithm does not run 8 times slower. It will definitely be a win.

So perhaps what you are thinking of is to use Ryu-printf and then do some additional work?

**Comment by expnkx** [comment]
> Does Ryu-printf work only for limited length strings?
> 
> > I've tested ryu sto algorithm. It runs around 0.4s per 10M double. Microsoft charconv will run at least 3 sec (sometimes it will run 7 seconds is your exp part is large enough). So as long as the verification algorithm does not run 8 times slower. It will definitely be a win.
> 
> So perhaps what you are thinking of is to use Ryu-printf and then do some additional work?

Yea. It has a precision scientific mode and that is something should work since we can pick the length whatever we want.

BTW, I do not know whether we can do something better here since we might not create the real string. It can be compared in some middle process.

Unfortunately, I do not fully understand how ryu algorithm works. That is why I am asking here.

**Comment by jk-jeon** [comment]
>It has a precision scientific mode and that is something should work since we can pick the length whatever we want.

I still don't get it. What exactly is your goal? Does Ryu-printf only for strings with limited number of digits, so that you want an algorithm that works for arbitrary length strings by augmenting what Ryu-printf is doing? Or, Ryu-printf is already fine with arbitrary length strings, but you want to invent something faster than that?

**Comment by expnkx** [comment]
> I still don't get it. What exactly is your goal? Does Ryu-printf only for strings with limited number of digits, so that you want an algorithm that works for arbitrary length strings by augmenting what Ryu-printf is doing? Or, Ryu-printf is already fine with arbitrary length strings, but you want to invent something faster than that?

Nope. Not ryu-printf.

It is arbitrary length of decimals to floating-point. It is the reverse question.

Ryu can only solve length of decimals which lengths are smaller than 18

**Comment by jk-jeon** [comment]
> Edit: I'm sorry, for whatever reason I was misusing the term Ryu-printf. It should be Ryu-string-to-float.

Perhaps the proper name is Ryu-parse I guess.

Anyway, now I can understand what's your issue.

**Comment by expnkx** [comment]
> > Edit: I'm sorry, for whatever reason I was misusing the term Ryu-printf. It should be Ryu-string-to-float.
> 
> Perhaps the proper name is Ryu-parse I guess.
> 
> Anyway, now I can understand what's your issue.

In your paper, parse and printf and equally important. Unfortunately, the reverse problem (which is parsing) is an untouched place. It is not fully solved compared to ryu dtos or grisu-exact is doing here.

**Comment by expnkx** [comment]
BTW, adding trailing zeros after round-trip for general and fixed shortest is de facto the most correct answer here since it never introduces round-trip verification problem for the input side, it definitely runs much faster. It is not cheating tbh from the middle point of view since it does not introduce problems compared to other digits for the input side.

**Comment by jk-jeon** [comment]
It seems that the problem is more complicated than I initially thought, and it requires thorough investigation, which I'm not sure if I can afford time for doing.

My guess is that Ulf is already working on this problem and if that's the case he will come up with a solution pretty soon.

>BTW, for outputting, adding trailing zeros after round-trip for general and fixed shortest is de facto the most correct answer here since it never introduces round-trip verification problem for the input side, it definitely runs much faster. It is not cheating tbh from the middle point of view since it does not introduce problems compared to other digits for the input side.

You can do whatever you want with your library. And I said my opinion is that there is little reason to add trailing zeros. If you are working on shortest output routine (it can be fixed-point, scientific, or general), then just output what the algorithm spits out. If you are working on fixed-precision output routine (again it can be fixed-point, scientific, or general), use a different algorithm like Ryu-printf (which is not equal to Ryu).
